File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map <int, int> m;
        for (auto n: nums) m[n]++;
        vector<vector<int>> count(nums.size() + 1);
        for (auto it: m){
            count[it.second].push_back(it.first);
        }
        vector<int> ans;
        
        for (int i=count.size() -1; i >= 0; i--){
            if (ans.size() == k) break;
            for (auto n: count[i]){
                ans.push_back(n);
            }
        }
        
        return ans;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {

        Map<Integer, Integer> mp = new HashMap<>();

        for (int n:nums) {
            mp.put(n, mp.getOrDefault(n, 0) + 1);
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> Integer.compare(mp.get(a), mp.get(b)));

        for (Map.Entry<Integer, Integer> e: mp.entrySet()) {
            pq.add(e.getKey());
            if (pq.size() > k) pq.poll();
        }

        int[] res = new int[k];

        int idx = 0;

        while (!pq.isEmpty()) {
            res[idx++] = pq.poll();
        }

        return res;
        
    }
}